### PR TITLE
Allow transient range owner to serve as repair coordinator

### PR DESCRIPTION
**Pull request by ifesdjeen**
cc @bdeggleston @aweisberg

### CODE REVIEW

File: src/java/org/apache/cassandra/repair/RepairRunnable.java [hunk 1]
```diff
@@ -651,7 +651,7 @@ private static void addRangeToNeighbors(List<CommonRange> neighborRangeList, Ran
         for (CommonRange cr : neighborRangeList)
         {
             // Use strict equality here, as worst thing that can happen is we generate one more stream
-            if (Iterables.elementsEqual(cr.endpoints, endpoints) && Iterables.elementsEqual(cr.transEndpoints, transEndpoints))
+            if (Sets.difference(cr.endpoints, endpoints).isEmpty() && Sets.difference(cr.transEndpoints, transEndpoints).isEmpty())
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/repair/RepairRunnable.java]: You can use .equals() here. Set equality does the right thing.

Comment (ifesdjeen) [reply src/java/org/apache/cassandra/repair/RepairRunnable.java]: Right, `return s.size() == o.size() && s.containsAll(o);`, i should've did that from starters.

**Comment by bdeggleston** [comment]
I think I've mostly convinced myself this is all you need to do to support coordinating repairs for locally transient ranges. I have a commit [here](https://github.com/bdeggleston/cassandra/tree/alexp/tr-followup-1-updates) here with a few adjustments:

* Move set comparison logic in RepairRunnable#addRangeToNeighbors to CommonRange class 
* changed asserts to Preconditions
* added a SymmetricLocalSync task test for the local transient case
* misc formatting

This leaves the sync task class hierarchy in kind of a messy state. We have AsymmetricSyncTask, but we also have SymmetricSyncTask which can also function as an asymmetric sync task. I do think the state of things is preferable to adding more branches to RepairJob#standardSyncing though, and a refactor is out of scope for this ticket. I'll open a follow on jira to clean that up once this is committed.

**Comment by ifesdjeen** [comment]
@bdeggleston thank you for the review and changes, they look great!

I agree the class hierarchy became not very optimal. I was concentrating on the AssumetrciLocalSyncTask and didn't change the symmetric one. I've added a follow-up commit that makes only one of them necessary and renames it to LocalSyncTask. It did not seem necessary to do the same with remote tasks as there we do not depend on the argument order, which we do in the local case because of transiency. I did try to avoid this change but couldn't find a good way to make it order-independent, but I've changed argument names and made sure corresponding checks are in place.

File: src/java/org/apache/cassandra/repair/LocalSyncTask.java [hunk 2]
```diff
@@ -39,43 +40,60 @@
...
+                               RangesAtEndpoint.toDummyList(Collections.emptyList()), desc.columnFamily);
+        }
 
-        if (!pullRepair && !remoteIsTransient)
+        if (transferRanges)
         {
             // send ranges to the remote node if we are not performing a pull repair
             // see comment on RangesAtEndpoint.toDummyList for why we synthesize replicas here
-            plan.transferRanges(dst, desc.keyspace, RangesAtEndpoint.toDummyList(differences), desc.columnFamily);
+            plan.transferRanges(remote, desc.keyspace, RangesAtEndpoint.toDummyList(differences), desc.columnFamily);
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/repair/LocalSyncTask.java:96]: So the idiom where it returns this isn't always used to return "this" some of the time it's a copy. Granted here it is returning this, but should you assign the plan field on the stack again?

I don't feel strongly about it just noticing that.

File: src/java/org/apache/cassandra/repair/RepairJob.java [hunk 3]
```diff
@@ -165,107 +166,116 @@ private boolean isTransient(InetAddressAndPort ep)
...
                 {
-                    TreeResponse r2 = trees.get(j);
+                    TreeResponse self = r1.endpoint.equals(local) ? r1 : r2;
+                    TreeResponse remote = r2.endpoint.equals(local) ? r1 : r2;
 
-                    if (isTransient(r1.endpoint) && isTransient(r2.endpoint))
+                    // pull only if local is full
+                    boolean requestRanges = !isTransient(self.endpoint);
+                    // push only if remote is full; additionally check for pull repair
+                    boolean transfterRanges = !isTransient(remote.endpoint) && !session.pullRepair;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/repair/RepairJob.java]: Typo in transfer

File: src/java/org/apache/cassandra/service/ActiveRepairService.java [hunk 4]
```diff
@@ -20,10 +20,25 @@
 import java.io.IOException;
 import java.lang.management.ManagementFactory;
 import java.net.UnknownHostException;
-import java.util.*;
-import java.util.concurrent.*;
+import java.util.ArrayList;
```
Comment (aweisberg) [inline src/java/org/apache/cassandra/service/ActiveRepairService.java]: Import fighting still happening?

File: src/java/org/apache/cassandra/repair/LocalSyncTask.java (hunk 2 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/repair/LocalSyncTask.java:96]: I felt like this is a side-effect generating, imperative method, so ignoring its return is fine as it's done in order to facilitate chaining.

File: src/java/org/apache/cassandra/repair/RepairJob.java (hunk 3 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/repair/RepairJob.java]: Thanks for noticing!

File: src/java/org/apache/cassandra/service/ActiveRepairService.java (hunk 4 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/service/ActiveRepairService.java]: Yes, I'll remove it on commit when it's squashed.

**Comment by ifesdjeen** [comment]
@aweisberg fixed imports and spelling. Thank you for spotting those.

**Comment by bdeggleston** [comment]
@ifesdjeen did some further consolidation of the sync task class hierarchy [here](https://github.com/bdeggleston/cassandra/tree/alexp/tr-followup-1/review2). I'm not really opposed to doing this refactor here instead of a separate ticket, but @krummas should probably take a look now :)

**Comment by ifesdjeen** [comment]
Heh, now I need to take a look as well)

**Pull request closed without merging**